numpy/_core/numeric.py

```diff
@@ -2554,17 +2554,17 @@ def array_equal(a1, a2, equal_nan=False):
     if a1.shape != a2.shape:
         return False
     if not equal_nan:
-        return builtins.bool((a1 == a2).all())
-    cannot_have_nan = (_dtype_cannot_hold_nan(a1.dtype)
-                       and _dtype_cannot_hold_nan(a2.dtype))
-    if cannot_have_nan:
-        if a1 is a2:
-            return True
-        return builtins.bool((a1 == a2).all())
+        return builtins.bool((asarray(a1 == a2)).all())
```

Comment:
Note this changes behaviour for `MaskedArray`, since it returns a masked array of bool, and here you strip the mask. In principle, not bad conceptually, since the underlying value is `False` if one array element is masked and the other is not, and `True` if both are masked. But definitely a change in API that might catch people by surprise: it will no longer be just `(a1 == a2).all()`, which I think is what most people would expect.

Reply:
p.s. I did not check the reasoning for the `asarray` - `asanyarray` would solve the problem with `MaskedArray` - I think that may be good enough to solve the problem here too? Alternatively, replace `(...).all()` with `np.all(...)` - that will deal with python bools fine.

Reply:
Hmmm, maybe `np.all` is best, then.  The `asarray` was restoring old behavior, but hand't though about it potentially making a regression for masked (where removing it was maybe an improvement)...

Reply:
Using `asanyarray` preserves the masked array and solves issue https://github.com/numpy/numpy/issues/27271, so I updated the PR to use that.

@mhvk Does that address your point? As far as I understand it, the problem with `asarray` is that it creates an unnecessary copy of the masked array (and hence uses more memory). 

I am also looking into updating the scalar comparison to solve #27271 by making `np.array(1) == np.array('s')` return a numpy bool instead of a python bool.

Reply:
The problem is not the use of memory, but rather than `np.asarray` makes any array subclass into a regular `ndarray`. So, `np.asanyarray` solves that!